contracts/settling_game/L05_Wonders.cairo

```diff
@@ -0,0 +1,280 @@
+            treasury_address,
+            22, 
+            ids_arr, 
+            22, 
+            amounts_arr)
+    else:
+
+    # Recurse
+    return loop_epochs_claim(
+        caller,
+        token_id,
+        current_epoch,
+        claiming_epoch + 1)
+    end
+    return ()
+end
+
+func loop_resources_claim{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+            wonders_state_address : felt, token_id : Uint256, epoch : felt, epoch_total_wonders : felt,
+            resource_claim_ids_len : felt, 
+            resource_claim_ids : felt*, 
+            resource_claim_amounts_len : felt, 
+            resource_claim_amounts : felt*) 
+        -> (
+            resource_claim_ids_len : felt, 
+            resource_claim_ids : felt*, 
+            resource_claim_amounts_len : felt, 
+            resource_claim_amounts : felt*):
+    alloc_locals
+
+    let ( below_max_id ) = is_nn_le(resource_claim_ids_len, 22)
+
+    assert_in_range(resource_claim_amounts_len, 1, 2**15)
+    if below_max_id == 1:
+        let ( resource_pool ) = IS05_Wonders.get_tax_pool(wonders_state_address, epoch, resource_claim_ids_len)
+        assert resource_claim_ids[resource_claim_ids_len - 1] = resource_claim_ids_len
+        assert resource_claim_amounts[resource_claim_ids_len - 1] = resource_pool / epoch_total_wonders 
+    
+        return loop_resources_claim(
+            wonders_state_address, token_id, epoch, epoch_total_wonders,
+            resource_claim_ids_len + 1, 
+            resource_claim_ids, 
+            resource_claim_amounts_len + 1, 
+            resource_claim_amounts)
+    else:
+    return (
+        resource_claim_ids_len, 
+        resource_claim_ids, 
+        resource_claim_amounts_len, 
+        resource_claim_amounts)
+    end
+end
+
+# Called everytime a user settled, unsettles or claims taxes
+@external
+func update_epoch_pool{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    alloc_locals
+    let ( controller ) = controller_address.read()
+    let ( wonders_state_address ) = IModuleController.get_module_address(contract_address=controller, module_id=9)
+    let ( calculator_address ) = IModuleController.get_module_address(contract_address=controller, module_id=7)
+
+    let ( current_epoch ) = IL04_Calculator.calculateEpoch(calculator_address)
+    let ( last_updated_epoch ) = IS05_Wonders.get_last_updated_epoch(wonders_state_address)
+
+    let epochs_to_update = current_epoch - last_updated_epoch
+
+    if epochs_to_update == 0:
+        return ()
+    else:
+        let updating_epoch = current_epoch - epochs_to_update
+
+        let ( total_wonders_staked ) = IS05_Wonders.get_total_wonders_staked(wonders_state_address, updating_epoch) 
+        IS05_Wonders.set_total_wonders_staked(wonders_state_address, updating_epoch, total_wonders_staked) 
+        IS05_Wonders.set_last_updated_epoch(wonders_state_address, updating_epoch)
+
+        return update_epoch_pool()
```

Comment:
Why does this recursively call itself again?

Reply:
The function rolls over the total wonders staked for an epoch forward by one epoch, and it does this every recursion